specs/custom-iresourcemanager-spec.md

```diff
@@ -0,0 +1,101 @@
+
+# API Pages
+
+_(Each of the following L2 sections correspond to a page that will be on docs.microsoft.com)_
+
+## Application.ResourceManagerInitializing Event
+
+Raised during startup of a new WinUI thread to give the app a chance to provide its own [`IResourceManager`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.resources.iresourcemanager?view=windows-app-sdk-1.2)
+implementation to be used by the framework for resolving resource URIs.
+
+_Spec note: is there a version of the 'ResourceDictionary and XAML resource references' article that 
+links to the Windows App SDK documentation for relevant APIs?_
+
+```c#
+public event TypedEventHandler<object,ResourceManagerInitializingEventArgs> 
+ResourceManagerInitializing
+
+```
+
+### Remarks
+
+It is strongly recommended that apps which need to listen for this event register their handler in their `App` 
+class's constructor so that it is available during initial app launch. This event is raised once per WinUI
+thread during initialization. Because the framework assumes that each `IResourceManager` instance returned
+through the event is independent, it is your responsibility to ensure that the implementation is thread-safe
+if you wish to share the same instance between multiple threads.
+
+
+
+## ResourceManagerInitializingEventArgs Class
+
+Provides event data for the `Application.ResourceManagerInitializing` event.
+
+```c#
+public sealed class ResourceManagerInitializingEventArgs
+```
+
+### Remarks
+
+`ResourceManagerInitializingEventArgs` is used to provide the WinUI 3 framework with a custom implementation
+of [`IResourceManager`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.resources.iresourcemanager?view=windows-app-sdk-1.2)
+to be used for resolving resource URIs rather than the default `ResourceManager` instance that the framework
+creates. In the event handler, you should instantiate your custom `IResourceManager` and assign it to the
+`ResourceManagerInitializingEventArgs.ResourceManager` property. The value of this property is initially
+`null`, and it is only checked by the framework once per event raise after all registered event handlers have
+been invoked. If the property value is still `null` then the framework will use the default `ResourceManager`
+instance.
+
+
+## ResourceManagerInitializingEventArgs.ResourceManager Property
+
+Gets the explanation of the XAML resource reference failure.
+
+```c#
+public object ResourceManager { get; set; }
+```
+
+
+# API Details
+```c#
+namespace Microsoft.UI.Xaml
+{
+  runtimeclass Application
+  {
+    // existing ...
+
+    event Windows.Foundation.TypedEventHandler<object,Microsoft.UI.Xaml.ResourceManagerInitializingEventArgs> ResourceManagerInitializing;
+  };
+
+  runtimeclass ResourceManagerInitializingEventArgs
+  {
+    ResourceManagerInitializingEventArgs();
+    object ResourceManager { get; set; };
```

Comment:
```suggestion
    IResourceManager ResourceManager { get; set; };
```

Reply:
Unfortunately that breaks C#/WinRT interop assembly generation in the form of namespace collisions between `Windows.*` and MRT Core's `Microsoft.Windows.ApplicationModel.Resources` namespace.

Reply:
Can we work around by filtering the cswinrt projection to exclude Windows.ApplicationModel.Resources.* (or just ...IResourceManager)?

Reply:
The namespace collision isn't arising because in the process of generating the C#/WinRT interop assembly for WinUI we're also creating a projection for `IResourceManager`; the collision is happening because we're *referencing* the existing interop assembly that defines that projection.

Binlog and generated files: [binlog.zip](https://github.com/microsoft/microsoft-ui-xaml/files/10926883/binlog.zip)

As you can see from the attached archive's contents, none of the generated files for the interop assembly declare a namespace rooted by `Microsoft.Windows.` In fact, the only references to such a namespace lie in the usages of `Microsoft.Windows.ApplicationModel.Resources.IResourceManager` which are necessary if we want to strongly type the event args. And in order to resolve that type it's necessary to add `Microsoft.Windows.ApplicationModel.Resources.Projection.dll` as a reference which is unfortunately enough to trigger the namespace collision.

Reply:
seems like the tools should be fixed and not worked around in the API design.
if you must keep Object update the property name to include terms from the namespace and type so readers have some clue what this is.

Reply:
C#/WinRT 2.0.2 allows this to work.